ddtrace/internal/runtime/runtime_metrics.py

```diff
@@ -0,0 +1,112 @@
+import threading
+import time
+import itertools
+
+from ..logger import get_logger
+from .constants import (
+    DEFAULT_RUNTIME_METRICS,
+    DEFAULT_RUNTIME_TAGS,
+)
+from .metric_collectors import (
+    GCRuntimeMetricCollector,
+    PSUtilRuntimeMetricCollector,
+)
+from .tag_collectors import (
+    TracerTagCollector,
+    PlatformTagCollector,
+)
+
+log = get_logger(__name__)
+
+
+class RuntimeCollectorsIterable:
+    def __init__(self, enabled=None):
+        self._enabled = enabled or self.ENABLED
+        # Initialize the collectors.
+        self._collectors = [c() for c in self.COLLECTORS]
+
+    def __iter__(self):
+        collected = (
+            collector.collect(self._enabled)
+            for collector in self._collectors
+        )
+        return itertools.chain.from_iterable(collected)
+
+    def __repr__(self):
+        return '{}(enabled={})'.format(
+            self.__class__,
+            self._enabled,
+        )
+
+
+class RuntimeTags(RuntimeCollectorsIterable):
+    ENABLED = DEFAULT_RUNTIME_TAGS
+    COLLECTORS = [
+        TracerTagCollector,
+        PlatformTagCollector,
+    ]
+
+
+class RuntimeMetrics(RuntimeCollectorsIterable):
+    ENABLED = DEFAULT_RUNTIME_METRICS
+    COLLECTORS = [
+        GCRuntimeMetricCollector,
+        PSUtilRuntimeMetricCollector,
+    ]
+
+
+class RuntimeWorker(object):
+    FLUSH_INTERVAL = 10
+
+    def __init__(self, statsd_client, flush_interval=None):
+        self._lock = threading.Lock()
+        self._stay_alive = None
+        self._thread = None
+        self._flush_interval = flush_interval or self.FLUSH_INTERVAL
+        self._statsd_client = statsd_client
+        self._runtime_metrics = RuntimeMetrics()
+
+    def _target(self):
+        while True:
+            with self._lock:
+                if not self._stay_alive:
+                    break
+
+                self.flush()
+
+            time.sleep(self._flush_interval)
+
+    def start(self):
+        with self._lock:
+            if not self._thread:
+                log.debug("Starting {} thread".format(self))
+                self._thread = threading.Thread(target=self._target)
+                self._thread.setDaemon(True)
+                self._thread.start()
+                self._stay_alive = True
+
+    def stop(self):
+        with self._lock:
```

Comment:
I think like you could remove the lock and the `self._thread` check and just touch `_stay_alive`. That should be enough.

Reply:
@jd Removing the lock makes sense. Not sure about removing the check on `self._thread` -- would we still want `if not self._thread` here to not start new threads if someone calls `start()` multiple times on the same worker?

Reply:
I was only talking that check for `stop()`.

If you only toggle `self._stay_alive` in `stop`, then it does not matter if a thread exists or not; if the caller does:

Bad case `#1`:

```python
worker.stop()
worker.start()
```

The worker will just stop at the first iteration.

Bad case `#2` 
You call `stop()` multiple times: nothing bad happens.


I guess you don't want bad case `#1` to happen, so in that case you can keep the `self._thread` check. Though you don't need a lock in `stop()` because if there's a `self._thread`, then it's ok to stop anyway, and if there's no thread, you just return.

On another note, there's no way to call `start()` after `stop()` because; I don't know if it's an issue or not. You'd need to to set `self._thread = None` and in that case, you'd need the lock. Haha! 🤣